pkg/csi-common/nodeserver-default.go

```diff
@@ -71,6 +79,96 @@ func (ns *DefaultNodeServer) NodeGetCapabilities(ctx context.Context, req *csi.N
 }
 
 // NodeGetVolumeStats returns volume stats
-func (ns *DefaultNodeServer) NodeGetVolumeStats(ctx context.Context, in *csi.NodeGetVolumeStatsRequest) (*csi.NodeGetVolumeStatsResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "")
+func (ns *DefaultNodeServer) NodeGetVolumeStats(ctx context.Context, req *csi.NodeGetVolumeStatsRequest) (*csi.NodeGetVolumeStatsResponse, error) {
+
+	var err error
+	targetPath := req.GetVolumePath()
+	if targetPath == "" {
+		err = fmt.Errorf("targetpath %v is empty", targetPath)
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+	/*
+		volID := req.GetVolumeId()
+
+		TODO: Map the volumeID to the targetpath.
+
+		CephFS:
+		   we need secret to connect to the ceph cluster to get the volumeID from volume
+		   Name, however `secret` field/option is not available  in NodeGetVolumeStats spec,
+		   Below issue covers this request and once its available, we can do the validation
+		   as per the spec.
+
+		   https://github.com/container-storage-interface/spec/issues/371
+
+		RBD:
+		   Below issue covers this request for RBD and once its available, we can do the validation
+		   as per the spec.
+
+		   https://github.com/ceph/ceph-csi/issues/511
+
+	*/
+
+	isMnt, err := util.IsMountPoint(targetPath)
+
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, status.Errorf(codes.InvalidArgument, "targetpath %s doesnot exist", targetPath)
+		}
+		return nil, err
+	}
+	if !isMnt {
+		return nil, status.Errorf(codes.InvalidArgument, "targetpath %s is not mounted", targetPath)
+	}
+
+	cephfsProvider := volume.NewMetricsStatFS(targetPath)
+	volMetrics, volMetErr := cephfsProvider.GetMetrics()
```

Comment:
If it is a block volume what happens when we do a statfs on a file? Which stats are returned?

I would assume we need to check if this is a block volume and report total size.

Reply:
Block volume invocation is still under discussion from CO side, so its fine.

Reply:
Should we add a check and ignore the request if it is against a block volume, as we cannot depend on what the CO may do, and it would be prudent to stick to the specification?

Reply:
@ShyamsundarR As mentioned in earlier comment, its still under discussion at CO side or I dont see a risk here. I can take it as a follow up item or its well in my queue :)

Reply:
Filed https://github.com/ceph/ceph-csi/issues/519 as a tracker for the same, thanks.